phases/18-ethics-safety-alignment/03-direct-preference-optimization-family/code/main.py

```
@@ -0,0 +1,228 @@
+
+
+def main() -> None:
+    print("=" * 70)
+    print("DPO FAMILY ON TOY 4-ACTION PREFERENCE DATA (Phase 18, Lesson 3)")
+    print("=" * 70)
+    print(f"  true utility : {TRUE_UTILITY}")
+    print(f"  true optimum : {[f'{p:.3f}' for p in softmax(TRUE_UTILITY)]}")
+    print()
+
+    pairs = [sample_pref_pair() for _ in range(500)]
+    labels = [(random.randrange(N_ACTIONS),
```

Ruff (0.15.11)

[error] Standard pseudo-random generators are not suitable for cryptographic purposes

(S311)